plugins/plugin-java/che-plugin-java-plain/che-plugin-java-plain-server/src/main/java/org/eclipse/che/plugin/java/plain/server/inject/PlainJavaProjectSourceFolderWatcher.java

```diff
@@ -0,0 +1,144 @@
+import org.eclipse.lsp4j.FileChangeType;
+import org.eclipse.lsp4j.FileEvent;
+import org.eclipse.lsp4j.services.LanguageServer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Reports the create/update/delete changes on project source folders to jdt.ls
+ *
+ * @author V. Rubezhny
+ */
+public class PlainJavaProjectSourceFolderWatcher {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(PlainJavaProjectSourceFolderWatcher.class);
+
+  private static final Gson gson =
+      new GsonBuilder().disableHtmlEscaping().serializeNulls().create();
+
+  private final FileWatcherManager manager;
+  private final FileWatcherByPathMatcher matcher;
+  private final EventService eventService;
+
+  private final CopyOnWriteArrayList<Integer> watcherIds = new CopyOnWriteArrayList<>();
+
+  private LanguageServer languageServer;
+
+  @Inject
+  public PlainJavaProjectSourceFolderWatcher(
+      FileWatcherManager manager,
+      FileWatcherByPathMatcher matcher,
+      EventService eventService,
+      JavaLanguageServerExtensionService extensionService,
+      ProjectManager projectManager) {
+    this.manager = manager;
+    this.matcher = matcher;
+    this.eventService = eventService;
+  }
+
+  @PostConstruct
+  protected void startWatchers() {
+    int watcherId =
+        manager.registerByMatcher(
+            folderMatcher(),
+            s -> report(s, FileChangeType.Created),
+            s -> {},
+            s -> report(s, FileChangeType.Deleted));
+
+    watcherIds.add(watcherId);
+    eventService.subscribe(this::onServerInitialized, LanguageServerInitializedEvent.class);
+    eventService.subscribe(this::onProjectUpdated, ProjectUpdatedEvent.class);
+  }
+
+  @PreDestroy
+  public void stopWatchers() {
+    watcherIds.stream().forEach(id -> manager.unRegisterByMatcher(id));
+  }
+
+  private void onServerInitialized(LanguageServerInitializedEvent event) {
+    if (!JavaModule.LS_ID.equals(event.getId())) return;
+
+    LOG.debug(
+        "Registering source folder watching operations for language server : {}", event.getId());
+
+    languageServer = event.getLanguageServer();
+  }
+
+  private void onProjectUpdated(ProjectUpdatedEvent event) {
+    ExecuteCommandParams params =
+        new ExecuteCommandParams(
+            GET_PROJECT_SOURCE_LOCATIONS_COMMAND, singletonList(prefixURI(event.getProjectPath())));
+
+    languageServer
+        .getWorkspaceService()
+        .executeCommand(params)
+        .thenAccept(
+            result -> {
+              if (result == null) {
+                return;
+              }
+              Type type = new TypeToken<ArrayList<String>>() {}.getType();
+              List<String> paths = gson.fromJson(gson.toJson(result), type);
+              paths.stream().forEach(f -> matcher.accept(Paths.get(removeUriScheme(prefixURI(f)))));
+            });
+  }
+
+  private PathMatcher folderMatcher() {
+    return it -> isDirectory(it);
```

Comment:
We discussed this on one of dailies... I would not add any filters here, because to get the truly information on folder's project I have to ask jdt.ls to refresh the follder's project first, then ask the info from jdt.ls. Otherwise, the info on project might be incorrect. But if we do a refresh on project - why do we need to report a change on the folder? To refresh it once again?

For jdt.ls refreshing a workspace folder shouldn't be a costly operation in case the folder doesn't belong to any project. 

I tried to add here a filter based on project's source folders - it didn't work for some cases.

What if we're creating a folder (a package) inside a source folder in a moment when the project is not yet initialized properly? Your package would never be `visible` as a package from jds.ls point of view until it's parent is refreshed. So, you probably won't be able to move/copy a class into such a package. The problem is in very long project initialization - it takes more than a minute for me (when  https://github.com/eclipse/che/pull/11003 is applied)